File: dsa/DSA Spring '25/week1/twodigit.js
function twodigits(a, b) {
  let l = b.toString().length;
  let combinations = new Set(["2", "5"]);
  let res = new Set(["2", "5"]);

  if (b - a <= 10 && b <= 10) {
    if ((b < 2 && a < 2) || (a > 5 && b > 5)) {
      return 0;
    } else if (b >= 5 && a <= 2) {
      return combinations.length;
    } else if ((b < 5 && a >= 1) || b >= 2) {
      return 1;
    }
  }

  let idx = 2;

  while (idx <= l) {
    for (let curr of combinations) {
      let new_num2 = curr.padEnd(idx, 2);
      let new_num5 = curr.padEnd(idx, 5);

      if (new_num2 >= a && new_num2 <= b) {
        res.add(new_num2);
      }

      if (new_num5 >= a && new_num5 <= b) {
        res.add(new_num5);
      }

      combinations.add(new_num2);
      combinations.add(new_num5);
    }

    idx++;
  }

  if (a > 5) {
    return res.size - 2;
  } else if (a > 2) return res.size - 1;
  else return res.size;
}

console.log(twodigits(6, 936)); // 6
console.log(twodigits(60, 70)); // 0
console.log(twodigits(25, 25)); // 1
console.log(twodigits(1, 10 ** 9)); // 1022
console.log(twodigits(123456789, 987654321)); // 512

File: dsa/DSA Spring '25/week1/twodigit.py
def count_numbers(a, b):
    res = set(("2", "5"));
    combo = set(("2", "5"));
    
    if (b-a <= 10) and (b <= 10):
        if (b < 2 and a < 2) or (a > 5 and b > 5):
            return 0;
        elif (b >= 5 and a <= 2):
            return len(res);
        elif (b < 5 and a >= 1):
            return 1;

    blen = len(str(b))
    idx = 2;

    while idx <= blen:
        temp = set()
        for curr in combo:
            new_num2 = curr.ljust(idx, "2");
            new_num5 = curr.ljust(idx, "5");
            if (int(new_num2) >= a and int(new_num2) <= b):
                res.add(new_num2);
            if (int(new_num5) >= a and int(new_num5) <= b):
                res.add(new_num5);

            temp.add(new_num2)
            temp.add(new_num5)

        combo |= temp;
        
        idx += 1

    if a > 5: return len(res) - 2
    elif a > 2: return len(res) - 1
    else: return len(res)


